LFS-252: As a user, I can access the complete list of resources matching a query string that I entered in the top search bar

veronikaslc opened the pull request:
Functional for testing.
ToDo: move quick search results resource page to the separate hidden extension point

Commit:
LFS-252: As a user, I can access the complete list of resources matching a query string that I entered in the top search bar
ToDo: move quick search results resource page to the separate hidden extension point

sashaandjic commented:
I cannot see subjects in my search results.

marta- commented:
That's expected, in this version we still only search forms

sashaandjic commented:
tested

sdumitriu reviewed (changes requested):
One JS warning when there are more than one match in the same document: the key should be unique.

Should we use the answer UUID instead of the form one as the key for the row?

Or should we only show one match for each form instead?

I think having more than one hit per form is good, even though verbose, so I would go with the first fix.

OrdiNeu committed:
LFS-252: As a user, I can access the complete list of resources matching
a query string that I entered in the top search bar
Add a hidden extension point for components that are in the router but not
inside the sidebar

Commit:
LFS-252: As a user, I can access the complete list of resources matching a query string that I entered in the top search bar
Fixed same keys error; Made first colimn in a table as a link, added default of 10 rows, added scroll to midle setting

veronikaslc commented:
Fixed JS warning. I made a first column of search results table clickable link that invokes React redirect function `onClick` (exactly same way as it works for the quick search results dropdown). Somehow redirection from table page first tries to load url with hash, but then something just cuts hash off and loads the url without it. If you click on any of search results links in the first column, then after redirect in the browser click "go back", it will load a right url with hash and highlight the right question section. Just curious how's that possible.

sashaandjic commented:
Two minor things:
- when I want to go back to the search results page I have to click two times on the <- button 
- it would be nice if after clicking on the item from the search list the form opens and takes me to the field that contains the result. Currently the form opens up but it does not take me to the field shown in the search result.
![image](https://user-images.githubusercontent.com/23243244/92942333-65f5ea00-f41f-11ea-88ab-58841c11f4cf.png)

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/QuickSearchResults.jsx:
```diff
@@ -0,0 +1,123 @@
+    {
+      "key": "",
+      "label": "Identifier",
+      "format": (resultData) => (<Link href="#" onClick={() => {defaultRedirect(resultData, props)}}>{resultData.subject?.identifier || resultData["@name"] || anchor}</Link>),
```
The browser goes to `#/some/path` then to `#` because this is a link. The sequence of actions is:

- user clicks on the link
- the `onClick` handler runs and sets the location to the desired path
- the browser follows the link and sets the location to `#`

There are three possible fixes:
- use the right path for the `href` directly, without a custom onclick listener
- add a [`preventDefault`](https://developer.mozilla.org/en-US/docs/Web/API/Event/preventDefault) in the listener that will stop the browser from following the link
- remove the `href` completely

The best is the first, since that would be a proper link that can be middle-clicket to open in a new tab, and it uses less JS code hogging the memory.

acrowthe committed:
LFS-252: As a user, I can access the complete list of resources matching a query string that I entered in the top search bar
Improve search result link behaviour

sdumitriu committed:
LFS-252: As a user, I can access the complete list of resources matching a query string that I entered in the top search bar
Switch the hidden extension to a View

sdumitriu committed:
LFS-252: As a user, I can access the complete list of resources matching a query string that I entered in the top search bar
Removed no longer needed new extension point

sdumitriu committed:
LFS-252: As a user, I can access the complete list of resources matching a query string that I entered in the top search bar
Use the react-router Link instead of the MaterialUI one

sdumitriu commented on modules/commons/src/main/frontend/src/SearchBar.jsx:
```diff
@@ -57,6 +53,7 @@ function SearchBar(props) {
   const { classes, className, defaultValue, invertColors, onChange, onPopperClose, onSelect, onSelectFinish, queryConstructor, resultConstructor, staticContext, ...rest } = props;
   const [ search, setSearch ] = useState(defaultValue);
   const [ results, setResults ] = useState([]);
+  const [ moreResults, setMoreResults ] = useState([]);
```
We set a number later, and we check its value with `> 0`, so I think this is more correct:

```suggestion
  const [ moreResults, setMoreResults ] = useState(0);
```

sdumitriu commented on modules/commons/src/main/frontend/src/SearchBar.jsx:
```diff
@@ -240,6 +241,15 @@ function SearchBar(props) {
+                      className={classes.dropdownItem}
+                      key="more"
+                    >
+                    <Link to={"/content.html/QuickSearchResults#" + search} variant="body2" className={classes.root}>
```
The `Link` from MUI reloads the whole page, but the one from `react-router-dom` doesn't have a `variant` prop. Either:

- redefine all the needed style
- import the mui link as well, and use it as the [component](https://github.com/ReactTraining/react-router/blob/master/packages/react-router-dom/docs/api/Link.md#component-reactcomponent).

sdumitriu replied:
Using MuiLink directly doesn't work, there's JS errors and it reloads the page again.

sdumitriu reviewed (changes requested):
The quick search results popper doesn't close anymore.

sdumitriu commented on modules/commons/src/main/frontend/src/SearchBar.jsx:
```diff
@@ -240,6 +241,15 @@ function SearchBar(props) {
+                  { moreResults > 0 &&
+                    <MenuItem
+                      className={classes.dropdownItem}
+                      key="more"
```
Maybe this will fix the popper-not-closing issue:

```suggestion
                      key="more"
                      onClick={() => setPopperOpen(false)}
```

sashaandjic commented:
tested. all issues that I reported are fixed. 

I found one small issue that is probably out of scope. Here it is:
- user searches for "44" and gets the list of results
![image](https://user-images.githubusercontent.com/23243244/93653578-7aba1b00-f9e7-11ea-9638-70b03190437f.png)
- user clicks on the entry where the arrow points and gets this
![image](https://user-images.githubusercontent.com/23243244/93653592-8dcceb00-f9e7-11ea-9443-756e213d4c37.png)
- now user clicks on the Back bttn and gets this
![image](https://user-images.githubusercontent.com/23243244/93653630-bce35c80-f9e7-11ea-8178-c1028dc94cdf.png)
- if user clicks on the magnifying glass icon, this happens
![image](https://user-images.githubusercontent.com/23243244/93653649-d08ec300-f9e7-11ea-8aba-d636aaedde4a.png)

Also, it would be nice to be able to have links here:
![image](https://user-images.githubusercontent.com/23243244/93653691-00d66180-f9e8-11ea-9917-a154d5cd1aaf.png)

I can log both of my complains as separate issues.

sdumitriu reviewed (changes requested):
"X more results" is buggy:
- doesn't go away when closing and reopening the search bar
- doesn't get updated when the new string has less than 10 results

sdumitriu committed:
LFS-252: As a user, I can access the complete list of resources matching a query string that I entered in the top search bar
Bugfix: the "X more results" message doesn't disappear when it should

sdumitriu committed:
LFS-252: As a user, I can access the complete list of resources matching a query string that I entered in the top search bar
Use the right initial value for `moreResults`

sdumitriu committed:
LFS-252: As a user, I can access the complete list of resources matching a query string that I entered in the top search bar
Bugfix: quick search results dialog stays open after clicking on "X more results"

sdumitriu committed:
LFS-252: As a user, I can access the complete list of resources matching a query string that I entered in the top search bar
Bugfix: after clicking on a quick search result, JS error is logged and focusing the search bar again will return "no results"

sdumitriu committed:
LFS-252: As a user, I can access the complete list of resources matching a query string that I entered in the top search bar
Removed the `variant` prop that is only valid for MUI Link, but not for Router Link

sdumitriu reviewed (approved):
All bugs should be fixed, the only thing still suboptimal is the styling of the `X more results` link.

acrowthe reviewed (changes requested):
While on the search results page for a search term, searching for a new term and selecting view more results does not update the table of results
![image](https://user-images.githubusercontent.com/20056751/93776860-0c957400-fbf2-11ea-919c-afc60623615e.png)
![image](https://user-images.githubusercontent.com/20056751/93776955-2b940600-fbf2-11ea-83f8-26782f21cfd0.png)
![image](https://user-images.githubusercontent.com/20056751/93776993-33ec4100-fbf2-11ea-9988-87c8dd08766d.png)

sashaandjic commented:
Confirming problem reported by @acrowthe .
Previously reported problems are now fixed.

acrowthe committed:
LFS-252: As a user, I can access the complete list of resources matching a query string that I entered in the top search bar
Refresh results when new search entered from search results page

acrowthe reviewed (approved):
Refreshing issue fixed

sashaandjic commented:
tested. all good.

sashaandjic commented:
I would like to suggest the following type, title and order of columns for the result page:
**Identifier      Questionnaire                     Subject         Created by         Created on              Match**
-----
link                  Demographic data form     link               user name         date                          xxxxxxxxxxxx

so it looks similar to the other grids in the system
![image](https://user-images.githubusercontent.com/23243244/93808184-a3c1f200-fc19-11ea-829b-85d2ebf4855e.png)

The current look is a bit different and has a bit confusing titles


![image](https://user-images.githubusercontent.com/23243244/93808409-edaad800-fc19-11ea-8839-fb84193fdd51.png)

acrowthe commented:
@sashaandjic what are you expecting to be displayed in the `Subject` column? Because forms have no user friendly id, the displayed identifier is already the form's subject.
![image](https://user-images.githubusercontent.com/20056751/93811491-382e5380-fc1e-11ea-82ea-e68511c28094.png)

Adjusting the identifier column to match the forms actual identifier would look something like this:
![image](https://user-images.githubusercontent.com/20056751/93811694-89d6de00-fc1e-11ea-8b12-6b46f58e9f96.png)

acrowthe committed:
LFS-252: As a user, I can access the complete list of resources matching a query string that I entered in the top search bar
Adjust results columns to match other tables

acrowthe commented:
Columns adjusted
![image](https://user-images.githubusercontent.com/20056751/93815798-93634480-fc24-11ea-84c3-597225120a66.png)

marta- commented:
Now that I see the latest screenshot, please remove the Identifier column (users don't need to see that) and make Subject the first column.

sdumitriu commented:
That will work fine for forms, but once we enable searching in subjects and other things, it wouldn't work anymore.

marta- commented:
> That will work fine for forms, but once we enable searching in subjects and other things, it wouldn't work anymore.

Fair enough. I will create an issue to rethink the table once we actually do add other resources.

sashaandjic commented:
Had to build witn -Pquick option. The common columns for subject, forms, tumors and patients are:
- Identifier
- Created by
- Created on

Forms have user unfriendly identifier, however we can consider using the forms title ( e.g. patient123: demographic data). 
![image](https://user-images.githubusercontent.com/23243244/93834830-3b8c0400-fc4b-11ea-8ff4-eb75e1d6d8b9.png)
The whole thing may be shown as a link so user can open the form from the results page.

For all others this rule can be applied as well. E.g.  for tumors is Tumor: <name>; for patients Patient: <subject>
![image](https://user-images.githubusercontent.com/23243244/93834795-1f886280-fc4b-11ea-8200-7f8cf38c0898.png)

![image](https://user-images.githubusercontent.com/23243244/93834343-8c9af880-fc49-11ea-9298-ba3ca102d41a.png)

sashaandjic commented:
> Now that I see the latest screenshot, please remove the Identifier column (users don't need to see that) and make Subject the first column.

Without Identifier there is no way to open form/subject/tumor from the results page. Can we mask the ugly form id with something more user friendly (<subject>:<questionarrie> e.g. **Mary: Demographic data** as shown in the title of the form)

marta- commented:
And these column names will be generic enough for other resource types. Sounds good.

acrowthe committed:
LFS-252: As a user, I can access the complete list of resources matching a query string that I entered in the top search bar
Adjust results table columns and identifier text

acrowthe commented:
Column headers and identifier text adjusted as requested:
![image](https://user-images.githubusercontent.com/20056751/93899050-e939fa00-fcc1-11ea-888e-f065a894fc9f.png)

sdumitriu commented on modules/commons/src/main/frontend/src/headerStyle.jsx:
```diff
@@ -126,7 +126,8 @@ const headerStyle = theme => ({
     padding: "1px 2px",
-    borderRadius: "4px"
+    borderRadius: "4px",
+    marginLeft: "2px"
```
Why this change?

sdumitriu replied:
This change is here to add a space between at `QUICK SEARCH RESULTS FOR__ABC`. But it is wrong, since it adds unneeded space before the highlighted text match, which sometimes is within a word. The root problem is that the `QUICK SEARCH RESULTS FOR ABC` is MUI button, which has `display:flex` set to it, which removes whitespace space between its elements.

A better fix is to not use the same class for both match highlighting and the search page title.

OrdiNeu replied:
After trying a few things, I've decided to just swap the `display:flex` on the title for `display:inline-block`, which fixes the issue, and I've removed the `marginLeft`.

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/LiveTable.jsx:
```diff
@@ -74,12 +74,23 @@ function LiveTable(props) {
+
+  let triggerFetch = () => {
+    setFetchStatus(Object.assign({}, fetchStatus, {
+      "currentRequestNumber": -1,
```
I think this is wrong and causes the out-of-order results bug that Sasha mentioned.

sashaandjic replied:
Problem logged in LFS-578 - Out-of-order search results observed.

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/LiveTable.jsx:
```diff
@@ -351,7 +363,7 @@ function LiveTable(props) {
             :
             tableData ?
-              ( tableData.map(makeRow) )
+              ( tableData.map((item, index) => makeRow(item, index)) )
```
I don't think this is needed, the original type of call would also pass the two parameters (three actually, since the whole collection is passed as well).

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/QuickSearchResults.jsx:
```diff
@@ -0,0 +1,133 @@
+//
```
I think that this file should be in a different module, not in `data-entry`, since it is not tied to data entry, but the whole UI.

veronikaslc replied:
Should be moved to `commons` I guess.

OrdiNeu replied:
I'm moving it to `homepage`, since it seems to fit in best there. `commons` doesn't generally define endpoints.

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/QuickSearchResults.jsx:
```diff
@@ -0,0 +1,133 @@
+function QuickSearchResults(props) {
+
+  const { classes } = props;
+  // fix issue with classes
```
What issue? Leftover comment perhaps?

veronikaslc replied:
This line was inherited from `Subjects.jsx` that was taken as an example of page. If the `classes` is a problem there, it is an issue in `QuickSearchResults.jsx` as well as they use classes identically. On my opinion should be removed all together from both places as I do not see any issues with classes.

sashaandjic commented:
From the end user point of view, everything looks good. 
![image](https://user-images.githubusercontent.com/23243244/93938913-0093da00-fcf8-11ea-80d2-53af55dea8e6.png)

OrdiNeu committed:
LFS-252: As a user, I can access the complete list of resources matching
a query string that I entered in the top search bar
Apply suggestions from code review

OrdiNeu committed:
LFS-252: LFS-252: As a user, I can access the complete list of resources
matching
Move the link on search results outside of the menu item, so that clicking
anywhere on the menu item will redirect you to the extra results

OrdiNeu committed:
LFS-252: As a user, I can access the complete list of resources matching
a query string that I entered in the top search bar
Remove the left padding from highlighted text and instead alter the
styling of MUI buttons for the title of the Quick Search Results page

sdumitriu commented on modules/commons/src/main/frontend/src/SearchBar.jsx:
```diff
@@ -242,15 +242,15 @@ function SearchBar(props) {
                     </MenuItem>
                   ))}
                   { moreResults > 0 &&
+                  <Link to={"/content.html/QuickSearchResults#" + search} className={classes.root}>
```
Huh, I was expecting `MenuList` to only accept `MenuItem` as children.

sdumitriu reviewed (approved):
LFS-581 is causing minor issues, but it isn't caused by this PR.
